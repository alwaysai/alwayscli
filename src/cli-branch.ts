import { CliBranch, ExcludeCommandType } from './types';
import { BRANCH } from './constants';

type Config = ExcludeCommandType<CliBranch>;

export function CliBranch(config: Config) {
  const branch: CliBranch = {
    ...config,
    commandType: BRANCH,
  };
  return branch;
}
