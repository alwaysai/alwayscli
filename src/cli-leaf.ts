import { AnyNamedInputs, CliLeaf, ExcludeCommandType, AnyInput, CliInput } from './types';
import { LEAF } from './constants';

export function CliLeaf<
  TPositional extends AnyInput = CliInput<undefined, false>,
  TNamed extends AnyNamedInputs = {},
  TEscaped extends AnyInput = CliInput<undefined, false>
>(config: ExcludeCommandType<CliLeaf<TPositional, TNamed, TEscaped>>) {
  const cliLeaf: CliLeaf<TPositional, TNamed, TEscaped> = {
    ...config,
    commandType: LEAF,
  };
  return cliLeaf;
}
