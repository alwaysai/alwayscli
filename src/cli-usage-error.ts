export const CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';

export class CliUsageError extends Error {
  public readonly code: typeof CLI_USAGE_ERROR;
  constructor(message?: string) {
    super(message);
    this.code = CLI_USAGE_ERROR;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
