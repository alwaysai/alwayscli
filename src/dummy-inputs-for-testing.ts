import { Input } from './types';

export const DUMMY_INPUT_UNDEFINED_WAS_PASSED = 'undefined was passed';
export const DUMMY_INPUT_EMPTY_ARRAY_WAS_PASSED = 'undefined was passed';
export const DUMMY_INPUT_THROWN_INTENTIONALLY = 'thrown intentionally';
export const DUMMY_INPUT_THROW = 'throw';
export const DUMMY_INPUT_THROW_NON_TRUTHY = 'throw-non-truthy';

export const dummyInput: Input<string, false> = {
  placeholder: '',
  getValue(argv) {
    if (typeof argv === 'undefined') {
      return DUMMY_INPUT_UNDEFINED_WAS_PASSED;
    }
    if (argv.length === 0) {
      return DUMMY_INPUT_EMPTY_ARRAY_WAS_PASSED;
    }
    if (argv[0] === DUMMY_INPUT_THROW) {
      throw new Error(DUMMY_INPUT_THROWN_INTENTIONALLY);
    }
    if (argv[0] === DUMMY_INPUT_THROW_NON_TRUTHY) {
      throw '';
    }
    return argv[0];
  },
  getDescription: () => '',
};

export const dummyRequiredInput: Input<string, true> = {
  placeholder: '<foo>',
  required: true,
  getValue: dummyInput.getValue,
  getDescription: () => '',
};
