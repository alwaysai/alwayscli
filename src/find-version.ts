import readPkgUp = require('read-pkg-up');
import { dirname } from 'path';

export async function findVersion() {
  const mainModule = require.main;
  if (!mainModule) {
    // We not sure under what circumstances require.main would be undefined but
    // the NodeJS types say it might be.
    return undefined;
  }

  const found = await readPkgUp({
    cwd: dirname(mainModule.filename),
    normalize: false,
  });

  if (!found) {
    return undefined;
  }

  if (!found.packageJson.version) {
    return undefined;
  }
  return found.packageJson.version;
}
