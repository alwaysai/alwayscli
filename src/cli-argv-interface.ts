import { AnyCommand } from './types';
import { accumulateCommandStack } from './accumulate-command-stack';
import { accumulateArgvs } from './accumulate-argvs';
import { accumulateNamedValues } from './accumulate-named-values';
import { CliUsageError } from './cli-usage-error';
import { CliTerseError } from './cli-terse-error';
import { CLI_BRANCH } from './constants';
import { findVersion } from './find-version';
import { callGetValue } from './call-get-value';
import { LastCommand } from './last-command';

type ArgvInterface = (...argv: string[]) => Promise<any>;

export type CliEnhancer = (argvInterface: ArgvInterface) => ArgvInterface;

export function CliArgvInterface(
  rootCommand: AnyCommand,
  options: Partial<{ enhancer: CliEnhancer }> = {},
) {
  const { enhancer } = options;

  if (enhancer) {
    return enhancer(argvInterface);
  }

  return argvInterface;

  async function argvInterface(...argv: string[]) {
    if (['-v', '--version'].includes(argv[0])) {
      const version = await findVersion();
      if (!version) {
        throw new CliTerseError('Failed to find a "version" string');
      }
      return version;
    }

    const {
      foundHelp,
      commandNamesAndPositionalArgv,
      namedArgvs,
      escapedArgv,
    } = accumulateArgvs(argv);

    const restCommandNamesAndPositionalArgv = accumulateCommandStack(
      rootCommand,
      commandNamesAndPositionalArgv,
    );

    if (foundHelp) {
      // E.g.:
      //   cli branch0 branch1 --help
      // Same as:
      //   cli --help branch0 branch1
      throw new CliUsageError();
    }

    const lastCommand = LastCommand(rootCommand);

    if (lastCommand.commandType === CLI_BRANCH) {
      if (restCommandNamesAndPositionalArgv[0]) {
        // E.g. cli branch0 branch1 bad-command-name
        throw new CliUsageError(`Bad command "${restCommandNamesAndPositionalArgv[0]}"`);
      }
      // E.g. cli branch0 branch1
      throw new CliUsageError();
    }

    let argsValue: any = undefined;
    if (lastCommand.positionalInput) {
      // Note that for named and escaped argvs, we distinguish between
      // `undefined` and `[]`. For example, "cli" gives an escaped argv
      // `undefined` whereas "cli --" gives an escaped argv `[]`. For the
      // "positionalArgv", however, there is no such distinction. By convention,
      // we elect here to pass in `undefined` rather than an empty array when no
      // positional arguments are passed.
      const positionalArgv =
        restCommandNamesAndPositionalArgv.length > 0
          ? restCommandNamesAndPositionalArgv
          : undefined;
      argsValue = await callGetValue(lastCommand.positionalInput, positionalArgv);
    } else if (restCommandNamesAndPositionalArgv.length > 0) {
      throw new CliUsageError(
        `Unexpected argument "${restCommandNamesAndPositionalArgv[0]}" : Command "${
          lastCommand.name
        }" does not accept positional arguments`,
      );
    }

    const namedValues = await accumulateNamedValues(
      lastCommand.namedInputs || {},
      namedArgvs,
    );

    let escapedValue: any = undefined;
    if (lastCommand.escapedInput) {
      escapedValue = await callGetValue(lastCommand.escapedInput, escapedArgv, '--');
    } else if (escapedArgv) {
      throw new CliUsageError(
        `Command "${lastCommand.name}" does not allow "--" as an argument`,
      );
    }

    const result = await lastCommand.action(argsValue, namedValues, escapedValue);
    return result;
  }
}
