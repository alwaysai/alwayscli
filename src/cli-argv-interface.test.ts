import { runAndCatch } from '@carnesen/run-and-catch';
import { CliBranch } from './cli-branch';
import { CliLeaf } from './cli-leaf';
import { dummyInput } from './dummy-inputs-for-testing';
import { CliArgvInterface, CliEnhancer } from './cli-argv-interface';
import { findVersion } from './find-version';
import { CLI_USAGE_ERROR } from './cli-usage-error';

const leafWithNamedInputs = CliLeaf({
  name: 'leaf-with-named-inputs',
  namedInputs: {
    foo: dummyInput,
  },
  action(...args) {
    return args;
  },
});

const leafWithPositionalInput = CliLeaf({
  name: 'leaf-with-positional-input',
  positionalInput: dummyInput,
  action(...args) {
    return args;
  },
});

const leafWithEscapedInput = CliLeaf({
  name: 'leaf-with-escaped-input',
  escapedInput: dummyInput,
  action(...args) {
    return args;
  },
});

const root = CliBranch({
  name: 'cli',
  subcommands: [leafWithPositionalInput, leafWithNamedInputs, leafWithEscapedInput],
});

const argvInterface = CliArgvInterface(root);

describe(CliArgvInterface.name, () => {
  it('calls the enhancer if provided', async () => {
    const spy = jest.fn();
    const enhancer: CliEnhancer = argvInterface => async (...argv: string[]) => {
      spy(...argv);
      await argvInterface(...argv);
    };
    const enhancedArgvInterface = CliArgvInterface(leafWithPositionalInput, { enhancer });
    await enhancedArgvInterface('foo', 'bar');
    expect(spy.mock.calls).toEqual([['foo', 'bar']]);
  });
  it('returns version string from package.json if "-v" or "--version" is passed', async () => {
    const version = await findVersion();
    expect(await argvInterface('-v')).toBe(version);
    expect(await argvInterface('--version')).toBe(version);
  });

  it('throws USAGE error with empty message if --help is passed', async () => {
    const exception = await runAndCatch(argvInterface, '--help');
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toBeFalsy();
  });

  it('throws USAGE error with empty message if last command is a branch and no additional argv is present', async () => {
    const exception = await runAndCatch(argvInterface);
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toBeFalsy();
  });

  it('throws USAGE error "bad command" if last command is a branch and additional argv is present', async () => {
    const exception = await runAndCatch(argvInterface, 'oops');
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toMatch(/bad command/i);
    expect(exception.message).toMatch('"oops"');
    expect(exception.message).toMatchSnapshot();
  });

  it('throws USAGE error "positional arguments" if last command is a leaf without positionalInput property and additional argv is present', async () => {
    const exception = await runAndCatch(argvInterface, leafWithNamedInputs.name, 'oops');
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toMatch('Unexpected argument "oops"');
    expect(exception.message).toMatch(leafWithNamedInputs.name);
    expect(exception.message).toMatch(/positional arguments/i);
    expect(exception.message).toMatchSnapshot();
  });

  it('Passes parsed positional value as first argument of the "action" function', async () => {
    const positionalArgv = ['foo', 'bar'];
    const result = await argvInterface(leafWithPositionalInput.name, ...positionalArgv);
    expect(result).toEqual([dummyInput.getValue(positionalArgv), {}, undefined]);
  });

  it('Passes parsed named values as second argument of the "action" function', async () => {
    const namedArgv = ['--foo', 'bar'];
    const result = await argvInterface(leafWithNamedInputs.name, ...namedArgv);
    expect(result).toEqual([undefined, { foo: dummyInput.getValue(['bar']) }, undefined]);
  });

  it(`Throws USAGE error 'does not allow "--"' if leaf does not have an "escaped" property`, async () => {
    const exception = await runAndCatch(
      argvInterface,
      leafWithPositionalInput.name,
      '--',
    );
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toMatch(leafWithPositionalInput.name);
    expect(exception.message).toMatch('does not allow "--"');
  });

  it('Passes parsed escaped value as third argument of the "action" function', async () => {
    const result = await argvInterface(leafWithEscapedInput.name, '--');
    expect(result).toEqual([
      undefined,
      {},
      leafWithEscapedInput.escapedInput!.getValue([]),
    ]);
  });
});
