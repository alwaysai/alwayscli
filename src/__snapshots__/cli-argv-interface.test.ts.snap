// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CliArgvInterface throws USAGE error "bad command" if last command is a branch and additional argv is present 1`] = `"Bad command \\"oops\\""`;

exports[`CliArgvInterface throws USAGE error "positional arguments" if last command is a leaf without positionalInput property and additional argv is present 1`] = `"Unexpected argument \\"oops\\" : Command \\"leaf-with-named-inputs\\" does not accept positional arguments"`;
